File: problems/524_longest_word_in_dictionary_through_deleting/solution.py
from typing import List


class Solution:
    def findLongestWord(self, s: str, dictionary: List[str]) -> str:
        def is_subseq(w: str) -> bool:
            it = iter(s)
            return all(c in it for c in w)
        
        dictionary.sort(key=lambda x: (-len(x), x))
        for word in dictionary:
            if is_subseq(word):
                return word
        return ''

File: problems/524_longest_word_in_dictionary_through_deleting/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    string findLongestWord(string s, vector<string>& dictionary) {
        sort(dictionary.begin(), dictionary.end(), [](auto& a, auto& b) {
            return a.size() != b.size() ? a.size() > b.size() : a < b;
        });
        for (auto& word : dictionary) {
            int i = 0;
            for (int j = 0; j < (int)s.size() && i < (int)word.size(); j++)
                if (word[i] == s[j]) i++;
            if (i == (int)word.size()) return word;
        }
        return "";
    }
};
